crates/red_knot_python_semantic/src/types.rs

```diff
@@ -306,3 +312,359 @@ pub struct IntersectionType<'db> {
+                        builder.intersections.extend(sub.intersections);
+                        builder
+                    },
+                )
+        } else {
+            for inner in &mut self.intersections {
+                inner.add_positive(self.db, ty);
+            }
+            self
+        }
+    }
+
+    #[allow(dead_code)]
+    fn add_negative(mut self, ty: Type<'db>) -> Self {
+        if let Type::Union(union) = ty {
+            union
+                .elements(self.db)
+                .iter()
+                .map(|elem| self.clone().add_negative(*elem))
+                .fold(
+                    IntersectionTypeBuilder::empty(self.db),
+                    |mut builder, sub| {
+                        builder.intersections.extend(sub.intersections);
+                        builder
+                    },
+                )
+        } else {
+            for inner in &mut self.intersections {
+                inner.add_negative(self.db, ty);
+            }
+            self
+        }
+    }
+
+    #[allow(dead_code)]
+    fn build(self) -> Type<'db> {
+        let mut builder = UnionTypeBuilder::new(self.db);
+        for inner in self.intersections {
+            builder = builder.add(inner.build(self.db));
+        }
+        builder.build()
+    }
+}
+
+#[allow(unused)]
+#[derive(Debug, Clone)]
+struct InnerIntersectionTypeBuilder<'db> {
+    positive: FxOrderSet<Type<'db>>,
+    negative: FxOrderSet<Type<'db>>,
+}
+
+impl<'db> InnerIntersectionTypeBuilder<'db> {
+    fn new() -> Self {
+        Self {
+            positive: FxOrderSet::default(),
+            negative: FxOrderSet::default(),
+        }
+    }
+
+    /// Adds a positive type to this intersection.
+    fn add_positive(&mut self, db: &'db dyn Db, ty: Type<'db>) {
+        match ty {
+            Type::Intersection(inter) => {
+                let pos = inter.positive(db);
+                let neg = inter.negative(db);
+                self.positive.extend(pos.difference(&self.negative));
+                self.negative.extend(neg.difference(&self.positive));
+                self.positive.retain(|elem| !neg.contains(elem));
+                self.negative.retain(|elem| !pos.contains(elem));
+            }
+            _ => {
+                if !self.negative.remove(&ty) {
+                    self.positive.insert(ty);
+                };
+            }
+        }
+    }
+
+    /// Adds a negative type to this intersection.
+    fn add_negative(&mut self, db: &'db dyn Db, ty: Type<'db>) {
+        match ty {
+            Type::Intersection(inter) => {
+                let pos = inter.negative(db);
+                let neg = inter.positive(db);
+                self.positive.extend(pos.difference(&self.negative));
+                self.negative.extend(neg.difference(&self.positive));
+                self.positive.retain(|elem| !neg.contains(elem));
+                self.negative.retain(|elem| !pos.contains(elem));
```

Comment:
I assume you don't both with `.shrink_to_fit()` calls here because it'll probably cost more than it's worth for such a frequently called method, and you'll reclaim the memory in the `build()` method anyway. Is that right?

Reply:
Right, it only makes sense to `shrink_to_fit` when we are turning things immutable.